source/fuzz/fact_manager.cpp

```diff
@@ -696,9 +737,78 @@ void FactManager::DataSynonymAndIdEquationFacts::AddDataSynonymFactRecursive(
   assert(DataDescriptorsAreWellFormedAndComparable(context, dd1, dd2));
 
   // Record that the data descriptors provided in the fact are equivalent.
+  // Both |dd1| and |dd2| are registered in the equivalence relation after this
+  // point.
   MakeEquivalent(dd1, dd2);
 
-  // We now check whether this is a synonym about composite objects.  If it is,
+  // Compute various corollary facts.
+  ComputeConversionSynonymFacts(dd1, dd2, context);
+  ComputeCompositeDataSynonymFacts(dd1, dd2, context);
+}
+
+void FactManager::DataSynonymAndIdEquationFacts::ComputeConversionSynonymFacts(
+    const protobufs::DataDescriptor& dd1, const protobufs::DataDescriptor& dd2,
+    opt::IRContext* context) {
+  assert(synonymous_.Exists(dd1) && synonymous_.Exists(dd2) &&
+         "Descriptors should've been registered in the equivalence relation");
+
+  // The fact that |dd1| and |dd2| are synonymous allows us to produce more
+  // corollary facts.
+  assert(synonymous_.IsEquivalent(dd1, dd2) &&
+         "Descriptors shoul've been marked as synonymous");
+
+  const auto* representative = synonymous_.Find(&dd1);
+  assert(representative &&
+         "Representative can't be null for a registered descriptor");
+
+  const auto* type =
+      context->get_type_mgr()->GetType(fuzzerutil::WalkCompositeTypeIndices(
+          context, fuzzerutil::GetTypeId(context, representative->object()),
+          representative->index()));
+  assert(type && "Data descriptor has invalid type");
+
+  if ((type->AsVector() && type->AsVector()->element_type()->AsInteger()) ||
+      type->AsInteger()) {
+    // |dd1| and |dd2| are synonymous. If there exist equation facts of the form
+    // |%a = opcode %dd1| and |%b = opcode %dd2| where |opcode| is either
+    // OpConvertSToF or OpConvertUToF, then |a| and |b| are synonymous.
+    std::vector<const protobufs::DataDescriptor*> convert_s_to_f_lhs;
+    std::vector<const protobufs::DataDescriptor*> convert_u_to_f_lhs;
+
+    for (const auto& fact : id_equations_) {
+      for (const auto& equation : fact.second) {
+        if (synonymous_.IsEquivalent(*equation.operands[0], *representative)) {
+          if (equation.opcode == SpvOpConvertSToF) {
+            convert_s_to_f_lhs.push_back(fact.first);
+          } else if (equation.opcode == SpvOpConvertUToF) {
+            convert_u_to_f_lhs.push_back(fact.first);
+          }
+        }
+      }
+    }
+
+    for (const auto& synonym_vector :
```

Comment:
I would rather you changed `synonym_vector` to something like `synonyms`, or something else that doesn't include `vector`, to avoid possible confusion with the SPIR-V vector type (since the code above involves AsVector()).

If you think the fact that the data type is vector is important for readability then drop the use of `auto`.

Reply:
Fixed in d7d0f49.